feat: Stabilize MSRV-aware resolver config (#14639)
rust-lang/cargo: src/doc/src/reference/resolver.md

[semver trick]: https://github.com/dtolnay/semver-trick
[`downcast_ref`]: ../../std/any/trait.Any.html#method.downcast_ref

### Rust version

To support developing software with a minimum supported [Rust version],
the resolver can take into account a dependency version's compatibility with your Rust version.
This is controlled by the config field [`incompatible-rust-version`].

With the `fallback` setting, the resolver will prefer packages with a Rust version that is
equal to or greater than your own Rust version.
For example, you are using Rust 1.85 to develop the following package:
```toml
[package]
name = "my-cli"
rust-version = "1.62"

[dependencies]
clap = "4.0"  # resolves to 4.0.32
```

> The resolver would pick 4.0.32 as its Rust version is 1.60.0.
> - 4.0.0 also has a Rust version of 1.60.0 but that is a [lower version number](#version-numbers)
> - 4.5.20 is a much [higher version](#version-numbers) and it has a Rust version of 1.74.0 which is compatible with your 1.85 toolchain but it is incompatible with `my-cli`'s Rust version of 1.62

Reviewer: This feels more cohesive to me.

```suggestion
The resolver would pick version 4.0.32 for dependency `clap` because
- 4.0.32 has a Rust version of 1.60.0, which is compatible with `my-cli`'s Rust version of 1.62.
- 4.0.0 also has a Rust version of 1.60.0 but that is a [lower version number](#version-numbers).
- Despite 4.1.0 and later versions all have a much [higher version](#version-numbers), their Rust versions are incompatible with `my-cli`'s Rust vesion. For example, 4.5.20 has a Rust version of 1.74.0 which is compatible with your 1.85 toolchain, but not with `my-cli`'s Rust version of 1.62.
```

Author: Note that that is mixing sentence structures.  The header was changed to end with "because" and "because despite" doesn't make sense to say.  I'm working on my own teak.

Reviewer: Thanks! Looks nicer :)

If a version requirement does not include a Rust version compatible dependency version,
the resolver won't error but will instead pick a version, even if its potentially suboptimal.
For example, you change the dependency on `clap`:
```toml
[package]
name = "my-cli"
rust-version = "1.62"

[dependencies]
clap = "4.2"  # resolves to 4.5.20
```
No version of `clap` matches that [version requirement](#version-requirements)
that is compatible with Rust version 1.62.
The resolver will then pick an incompatible version, like 4.5.20 despite it having a Rust version of 1.74.

When the resolver selects a dependency version of a package,
it does not know all the workspace members that will eventually have a transitive dependency on that version
and so it cannot take into account only the Rust versions relevant for that dependency.
The resolver has heuristics to find a "good enough" solution when workspace members have different Rust versions.
This applies even for packages in a workspace without a Rust version.

> With multiple Rust versions, the resolver may pick a lower dependency version than necessary.

Reviewer: Make it explicit?

```suggestion
When a workspace has members with multiple different Rust versions,
the resolver may pick a lower dependency version than necessary.
```

or be clear we're talking about the above case?

```suggestion
In the above situation, the resolver may pick a lower dependency version than necessary.
```

For example, you have the following workspace members:
```toml
[package]
name = "a"
rust-version = "1.62"

[package]
name = "b"

[dependencies]
clap = "4.2"  # resolves to 4.5.20
```
Though package `b` does not have a Rust version and could use a higher version like 4.5.20,
4.0.32 will be selected because of package `a`'s Rust version of 1.62.

Or the resolver may pick too high of a version.
For example, you have the following workspace members:
```toml
[package]
name = "a"
rust-version = "1.62"

[dependencies]
clap = "4.2"  # resolves to 4.5.20

[package]
name = "b"

[dependencies]
clap = "4.5"  # resolves to 4.5.20
```
Though each package has a version requirement for `clap` that would meet its own Rust version,
because of [version unification](#version-numbers),
the resolver will need to pick one version that works in both cases and that would be a version like 4.5.20.

[Rust version]: rust-version.md

> [`incompatible-rust-version`]: config.md#resolverincompatible-rust-versions

Reviewer: ```suggestion
[`incompatible-rust-versions`]: config.md#resolverincompatible-rust-versions
```

[semver trick]: https://github.com/dtolnay/semver-trick
[`downcast_ref`]: ../../std/any/trait.Any.html#method.downcast_ref

### Rust version

To support developing software with a minimum supported [Rust version],
the resolver can take into account a dependency version's compatibility with your Rust version.
This is controlled by the config field [`resolver.incompatible-rust-versions`].

With the `fallback` setting, the resolver will prefer packages with a Rust version that is
equal to or greater than your own Rust version.
For example, you are using Rust 1.85 to develop the following package:
```toml
[package]
name = "my-cli"
rust-version = "1.62"

[dependencies]
clap = "4.0"  # resolves to 4.0.32

> `

Reviewer: Should be three backticks
